src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardStream.cs

```diff
@@ -0,0 +1,345 @@
+
+            _nativeWrapper = new NativeWrapper(_compressionMode, EnsureCompressionLevelIsValid(compressionLevel));
+            switch (_compressionMode)
+            {
+                case CompressionMode.Compress:
+                    _streamWriteHelper = new StreamWriteHelper(
+                        compressedStream: compressedStream,
+                        compressedBuffer: _arrayPool.Rent(_nativeWrapper.RecommendedOutputSize));
+                    break;
+                case CompressionMode.Decompress:
+                    _streamReadHelper = new StreamReadHelper(
+                        compressedStream: compressedStream,
+                        compressedBuffer: _arrayPool.Rent(_nativeWrapper.RecommendedInputSize));
+                    break;
+            }
+        }
+
+        public override bool CanRead => _compressedStream.CanRead &&
+                                        _compressionMode == CompressionMode.Decompress;
+
+        public override bool CanWrite => _compressedStream.CanWrite &&
+                                         _compressionMode == CompressionMode.Compress;
+
+        public override bool CanSeek => false;
+
+        public override long Length => throw new NotSupportedException();
+
+        public static int MaxCompressionLevel => NativeWrapper.MaxCompressionLevel; // maximum compression level available
+
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            if (!_disposed)
+            {
+                switch (_compressionMode)
+                {
+                    case CompressionMode.Compress:
+                        try
+                        {
+                            foreach (var outputBufferPosition in _nativeWrapper.FlushBySteps(_streamWriteHelper.CompressedBufferInfo))
+                            {
+                                _streamWriteHelper.WriteBufferToCompressedStream(count: outputBufferPosition);
+                            }
+                            _compressedStream.Flush();
+                        }
+                        finally
+                        {
+                            _nativeWrapper.Dispose();
+                            _arrayPool.Return(_streamWriteHelper.CompressedBufferInfo.Bytes);
+                        }
+                        break;
+                    case CompressionMode.Decompress:
+                        _nativeWrapper.Dispose();
+                        _arrayPool.Return(_streamReadHelper.CompressedBufferInfo.Bytes);
+                        break;
+                }
+
+                _disposed = true;
+            }
+        }
+
+        public override void Flush()
+        {
+            throw new NotSupportedException("Use Dispose instead.");
```

Comment:
This might be a mistake.

Consider that Dispose is NOT supposed to throw exceptions, and it seems likely that Flush methods *could* throw exceptions.

I think it would be more appropriate to have the flushing logic here and have the user of this stream call Flush before calling Dispose.

It's OK if Dispose ALSO calls Flush (and ignores exceptions in Flush), but that code path would only be as a safety net in case the user forgot to call Flush.

Reply:
Added. See these comments: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/94#discussion_r368568686